### PR TITLE
[Draft] low latency boundary condition generation

### PR DESCRIPTION
**Pull request by nicholasbalasus**
### Name and Institution (Required)

Name: Nick Balasus
Institution: Harvard University

### Describe the update

This draft PR will include updates to how we generate boundary conditions (likely `v2023-08` that will accompany the IMI 2.0). This includes some bug fixes but mostly changes that will allow us to generate boundary conditions with lower latency.

Changes that will impact the output:
- [x] switch to 2° x 2.5° (#146)
- [x] upgrade to GEOS-Chem v14.2.1 and ensure consistency with IMI (#147)
- [x] sample GEOS-Chem SpeciesConcVV_CH4 and LevelEdgeDiagn instantaneously instead of time-averaged (previous bug)
- [x] remove the dropping of outliers (#148)
- [x] use a restart file for 1 April 2018 from Todd's simulation (from NOAA surface data)
- [x] reduce the temporal smoothing of biases to -15 days (instead of +/- 15 days)
- [x] drop observations over water and south of 60°S
- [x] explore data density minimums (when regridding and when defining a zonal mean bias)
- [x] checking that our TROPOMI operator is behaving as expected (#110)

Changes that will not impact the output:
- [x] modify the default GEOS-Chem config files with `sed` instead of providing config files
- [x] add flexibility for number of vertical layers (#155)

### CODE REVIEW

**Comment by nicholasbalasus** [comment]
I am only requesting the review from @laestrada, but I will tag others in places they might want to chime in (thanks @djvaron, @msulprizio, @toddmooring, @eastjames, @megan-he!). This should close the five issues linked above.

This update uses GCClassic 14.2.1 to generate boundary conditions at 2.0° x 2.5°. They will be more appropriately named `v2023-09` (my branch name with `v2023-08` was too ambitious it seems). I will generate the full record (and compare to the 4.0° x 5.0° `v2023-06` boundary conditions) once this code looks good and 14.2.1 is officially released.

I have tested two things here:

**(1) The default IMI test case on Cannon.**

I tested this because GCClassic is now cloned as part of the setup for IMI (getting rid of the need for users to clone it themselves). This is the bash script used to test: [test-default-imi.txt](https://github.com/geoschem/integrated_methane_inversion/files/12561135/test-default-imi.txt). 

It works - results here if you're interested: [visualization_notebook.html.zip](https://github.com/geoschem/integrated_methane_inversion/files/12561477/visualization_notebook.html.zip)

**(2) The reproducibility of our boundary conditions generation.**

I generated three sets of boundary conditions: test 1 of 20180401-20180630, test 2 of 20180515-20180531 (using a restart generated by test 1), test 3 of 20180615-20180630 (using a restart generated by test 2). This is the bash script used to test: [test-bcs.txt](https://github.com/geoschem/integrated_methane_inversion/files/12561158/test-bcs.txt). The boundary conditions from test 2 and test 3 are consistent with those generated in test 1, as is tested with this simple Python code:

```
import xarray as xr
import numpy as np
from datetime import date, timedelta

basedir = "/n/holyscratch01/jacob_lab/nbalasus/"

test2_dates_to_check = [(date(2018, 5, 15) + timedelta(days=i)).strftime("%Y%m%d") for i in range(17)]
test3_dates_to_check = [(date(2018, 6, 15) + timedelta(days=i)).strftime("%Y%m%d") for i in range(16)]

for date in test2_dates_to_check:
    with xr.open_dataset(basedir + f"imi-test1/blended-boundary-conditions/GEOSChem.BoundaryConditions.{date}_0000z.nc4") as ds1,\
         xr.open_dataset(basedir + f"imi-test2/blended-boundary-conditions/GEOSChem.BoundaryConditions.{date}_0000z.nc4") as ds2:
        abs_diff = 1e9*np.abs((ds1["SpeciesBC_CH4"] - ds2["SpeciesBC_CH4"]).values)
        print(f"{date} -> {np.max(abs_diff)}")
        assert np.max(abs_diff) < 0.01, f"Diff is greater than 0.01 ppb for {date}"

for date in test3_dates_to_check:
    with xr.open_dataset(basedir + f"imi-test1/blended-boundary-conditions/GEOSChem.BoundaryConditions.{date}_0000z.nc4") as ds1,\
         xr.open_dataset(basedir + f"imi-test3/blended-boundary-conditions/GEOSChem.BoundaryConditions.{date}_0000z.nc4") as ds2:
        abs_diff = 1e9*np.abs((ds1["SpeciesBC_CH4"] - ds2["SpeciesBC_CH4"]).values)
        print(f"{date} -> {np.max(abs_diff)}")
        assert np.max(abs_diff) < 0.01, f"Diff is greater than 0.01 ppb for {date}"
 ```

File: src/inversion_scripts/operators/TROPOMI_operator.py [hunk 1]
```diff
@@ -323,6 +323,11 @@ def apply_tropomi_operator(
                 coords[1] + dlat / 2,
                 coords[1] + dlat / 2,
             ]
+            # If this is a global 2.0 x 2.5 grid, extend the eastern-most grid cells to 180 degrees
+            if (dlon == 2.5) & (coords[0] == 177.5):
+                for i in [1,2]:
+                    geoschem_corners_lon[i] += dlon / 2
+
```
Comment (nicholasbalasus) [inline src/inversion_scripts/operators/TROPOMI_operator.py:330]: @megan-he @eastjames this is not a big deal, just flagging it because it's kind of weird. To ensure assert abs(sum(overlap_area)-polygon_tropomi.area)/polygon_tropomi.area < 0.01 is true (that is, that we find a GEOS-Chem grid cell for all TROPOMI pixels which should be true on a global grid), I had to expand the eastern-most grid boxes on the 2.0 x 2.5 grid to 180 degrees (because the centers go from -180 to 177.5, so 178.75-180 did not exist!) Might be relevant elsewhere, but not a big deal.

File: src/inversion_scripts/utils.py [hunk 2]
```diff
@@ -293,6 +294,8 @@ def filter_tropomi(tropomi_data, xlim, ylim, startdate, enddate):
         & (tropomi_data["time"] <= enddate)
         & (tropomi_data["qa_value"] >= 0.5)
         & (tropomi_data["longitude_bounds"].ptp(axis=2) < 100)
+        & (tropomi_data["surface_classification"] != 1)
+        & (tropomi_data["latitude"] > -60)
```
Comment (nicholasbalasus) [inline src/inversion_scripts/utils.py:298]: @djvaron back in that long email thread about new boundary conditions, we decided to drop data over the water and south of 60S. This means it will be dropped in inversions as well (to be consistent with boundary conditions). Is that okay?

File: src/write_BCs/GEOSChem.Restart.20180401_0000z.nc4
Comment (nicholasbalasus) [inline src/write_BCs/GEOSChem.Restart.20180401_0000z.nc4:1]: @toddmooring I am using the file from `/n/jacob_lab/Users/tmooring/GCClassic-correcttropopause_archdirs/gc_2x25_merra2_CH4_ct_CLSBCr_prod1/Restarts/GEOSChem.Restart.20180401_0000z.nc4` - can you confirm that is correct?

File: src/inversion_scripts/operators/TROPOMI_operator.py (hunk 1 above)
Comment (megan-he) [reply src/inversion_scripts/operators/TROPOMI_operator.py:330]: Thanks Nick! I had to change the max longitude of my state vector from 180 to 177.5 previously to avoid index mismatching, so I think this should resolve that issue.

File: src/components/setup_component/setup.sh [hunk 3]
```diff
@@ -127,7 +127,15 @@ setup_imi() {
...
 
+    # Clone version 14.2.1 of GCClassic
     # Define path to GEOS-Chem run directory files
+    cd "${InversionPath}"
+    rm -rf GCClassic
+    git clone https://github.com/geoschem/GCClassic.git
+    cd GCClassic
+    git checkout dac5a54 # most recent dev/14.2.1 @ 1 Sep 2023 12:44 PM (update this once 14.2.1 officially released)
+    git submodule update --init --recursive
+    cd ..
```
[the remark this answers is not in the document]
Comment (laestrada) [reply src/components/setup_component/setup.sh:138]: Thanks Nick, I would prefer if there was a check to see if the GCClassic directory is present, rather than deleting and redownloading every time. This allows users more flexibility to modify GEOS-Chem. I think wrapping the above in `if [ ! -d "GCClassic" ]; then` would be sufficient.

File: src/write_BCs/write_boundary_conditions.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    nan_value_filler_3d = bias.copy() * np.nan
+    for i in range(len(LON)):
+        nan_value_filler_3d[:,:,i] = nan_value_filler_2d
+
+    # Use these values to fill NaNs
+    bias = bias.fillna(nan_value_filler_3d)
+
+    print(f"Average bias (GC-TROPOMI): {bias.mean().values:.2f} ppb\n")
+
+    # If there are still NaNs (this will happen when TROPOMI data is missing), use 0.0 ppb as the bias but warn the user
```
Comment (laestrada) [inline src/write_BCs/write_boundary_conditions.py:195]: Nice, this is a good idea

File: src/write_BCs/write_boundary_conditions.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    - average the TROPOMI XCH4 and GEOS-Chem XCH4 to a (2.0 x 2.5) grid for each day
+(2) Make a gridded (2.0 x 2.5 x daily) field of the bias between TROPOMI and GEOS-Chem
+    - subtract the TROPOMI and GEOS-Chem grids from part 1 to get a starting point for the bias
+    - smooth this field spatially (5 lon grid boxes, 5 lat grid boxes) then temporally (15 days backwards)
+    - fill NaN values with the latitudinal average at that time
+        - for a latitudinal average to be defined, there must be >= 30 grid cells at that latitude
+        - when a latitudinal average cannot be found, the closest latitudinal average is used
+(3) Write the boundary conditions
+    - using the bias from Part 2, subtract the (GC-TROPOMI) bias from the GC boundary conditions
+"""
```
Comment (laestrada) [inline src/write_BCs/write_boundary_conditions.py:46]: The content of this looks good. I do wonder if the main script is getting to be too bloated.. perhaps we should separate steps 1,2, and 3 into functions?

**Review by laestrada — CHANGES_REQUESTED** [review]
Overall looks good, some minor changes requested.

**Comment by nicholasbalasus** [comment]
Thanks, @laestrada! Everything is addressed now (except for `isRegional`, which will need to wait for now). Testing now.

File: src/inversion_scripts/operators/TROPOMI_operator.py (hunk 1 above)
Comment (eastjames) [reply src/inversion_scripts/operators/TROPOMI_operator.py:330]: Looks good @nicholasbalasus!

File: src/write_BCs/README.md [hunk 6]
```diff
@@ -11,6 +11,7 @@
    - `Partition`       - which partition to run the jobs on.
    - `restartFilePath` - restart file for GEOS-Chem.
       - if your simulation starts on 1 April 2018, this won't be used (`GEOSChem.Restart.20180401_0000z.nc4` will).
+         - this file comes from a simulation from a CH4 simulation by Todd Mooring that is constrained by NOAA surface observations.
```
Comment (laestrada) [inline src/write_BCs/README.md:14]: maybe remove the first from a simulation ;)

File: src/write_BCs/write_boundary_conditions.py [hunk 7]
```diff
@@ -233,4 +208,38 @@ def apply_tropomi_operator_to_one_tropomi_file(filename):
...
+        - fill NaN values with the latitudinal average at that time
+            - for a latitudinal average to be defined, there must be >= 30 grid cells at that latitude
+            - when a latitudinal average cannot be found, the closest latitudinal average is used
+    (3) Write the boundary conditions
+        - using the bias from Part 2, subtract the (GC-TROPOMI) bias from the GC boundary conditions
+    """
+
+    daily_means = part1(satelliteDir, start_time_of_interest, end_time_of_interest)
+    bias = part2(daily_means)
+    part3(bias)
```
Comment (laestrada) [inline src/write_BCs/write_boundary_conditions.py:245]: Can you make the function names descriptive? Something like: `create_daily_means()`, `calculate_bias()`, and `write_bias_corrected_files()`

**Comment by nicholasbalasus** [comment]
Thanks @laestrada! Let me know if this is ready to squash + merge.

**Review by laestrada — APPROVED** [review]
Thanks Nick, this looks good to me!

File: src/components/setup_component/setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (msulprizio) [reply src/components/setup_component/setup.sh]: @nicholasbalasus I'm not sure a warning would be good enough here because it might get overlooked. In order to ensure the correct version of GCClassic is used, we should check for the version number in `GCClassic/CMakeLists.txt`. In that file, you should see the text `VERSION 14.2.1`. If it doesn't match the expected version then the setup script can either stop with an error (along the lines of the message you have above) or preferably print a message telling the user GCClassic is being updated to X.Y.Z and automatically update the GCClassic repo and submodules.

Also, rather than checking out a specific commit, I would recommend using the version tag (e.g. `git checkout 14.1.1`). We don't recommend pegging it to a development version in general, but if you must then you should use the branch name instead of the commit ID (e.g. `git checkout dev/14.2.1`). We can change it to use the tag once the version is officially released/tagged.

**Comment by toddmooring** [comment]
Hi everyone but especially @nicholasbalasus and @laestrada . Apologies for the delayed response. I just want to make sure I understand how the boundary condition generation is supposed to work:

1) The user picks a start date for the inversion on or after 1 April 2018.
2) If the date is after 1 April 2018, global GEOS-Chem is run forward at 2x2.5/47L to the start date. The initial condition is a 1 April 2018 restart supplied by me, emissions are assumed based on some inventory.
3) From the start date on to the end date, the global 2x2.5/47L run writes boundary condition files.
4) The inversion is performed, using among other things the boundary conditions generated by the above process.

Assuming this is how it works, I have two questions/suggestions:

1) What happened to Daniel's suggestion of only doing a 1-month emissions-driven spinup after initializing with one of my restarts (see his 7/27/23 1:04 p.m. email)? Are we just now relying on the bias correction process (in which we make the GEOS-Chem state look more like TROPOMI) to mitigate biases created by emissions-driven spinups of more than 1 month?

2) Note that I actually now have monthly restarts as late as 1 May 2018 from a **2x2.5/47L** simulation. This simulation was initialized (from a 2x2.5/72L restart) at 1 May 2012, with that restart coming from a 2x2.5/72L simulation started on 1 January 1985. I think you should switch to using a native 47L restart, for two reasons: a) it skips the process of interpolating from 72L to 47L and b) the stratospheric methane climate has had ~6 years to adjust to the 47L structure. Probably not hugely different from the 72L structure, but why not get rid of this source of a potential fake transient if you can?

**Comment by toddmooring** [comment]
Just to be clear, I realize my description of how the boundary condition generation process works glosses over the application of bias corrections to GEOS-Chem fields to make them better match TROPOMI.

**Comment by nicholasbalasus** [comment]
Hi Todd - thanks for looking through this! 

The upshot of the code is GEOS-Chem (2x2.5/47L) is initialized with your 1 April 2018 restart file and run until present. The archived boundary conditions are corrected to look like TROPOMI (starting when the TROPOMI beings on 30 April 2018), which is what we need for nested-grid simulations for regional inversions. 

I think I understand how what you are referring to in (1) would work for making a restart file for a global simulation (that is, run 1 December 2018 with emissions starting with your restart and then correct the 1 January 2019 output with TROPOMI). However, I don't see how this would be practical for generating multiple years of boundary conditions that we need for nested-grid simulations.

For (2), I will switch to the 47 L file if you point me towards it. Thanks!

**Comment by toddmooring** [comment]
I have conveyed the 2x2.5/47L restart to @nicholasbalasus via Slack

**Comment by laestrada** [comment]
@toddmooring Thanks for providing the restart file! Just to further explain Nick's explanation -- we do not expect users to generate the boundary conditions themselves. Instead, we create an archive (available from the washu data portal and aws), which the IMI downloads for the relevant user dates.
